src/value.rs

```diff
@@ -0,0 +1,1613 @@
+            }
+            Value::I8(value) => Value::I8(value.wrapping_abs()),
+            Value::I16(value) => Value::I16(value.wrapping_abs()),
+            Value::I32(value) => Value::I32(value.wrapping_abs()),
+            Value::I64(value) => Value::I64(value.wrapping_abs()),
+            // f32/f64::abs() is not available in libcore
+            Value::F32(value) => Value::F32(if value < 0. { -value } else { value }),
+            Value::F64(value) => Value::F64(if value < 0. { -value } else { value }),
+            Value::U8(_) | Value::U16(_) | Value::U32(_) | Value::U64(_) => self,
+        };
+        Ok(value)
+    }
+
+    /// Perform a negation operation.
+    ///
+    /// If the value type is `Generic`, then it is interpreted as a signed value.
+    ///
+    /// This corresponds to the DWARF `DW_OP_neg` operation.
+    pub fn neg(self, addr_mask: u64) -> Result<Value> {
+        // wrapping_neg() can be used because DWARF specifies that the result is undefined
+        // for negative minimal values.
+        let value = match self {
+            Value::Generic(value) => {
+                Value::Generic(sign_extend(value, addr_mask).wrapping_neg() as u64)
+            }
+            Value::I8(value) => Value::I8(value.wrapping_neg()),
+            Value::I16(value) => Value::I16(value.wrapping_neg()),
+            Value::I32(value) => Value::I32(value.wrapping_neg()),
+            Value::I64(value) => Value::I64(value.wrapping_neg()),
+            Value::F32(value) => Value::F32(-value),
+            Value::F64(value) => Value::F64(-value),
+            // It's unclear if these should implicity convert to a signed value.
+            // For now, we don't support them.
+            Value::U8(_) | Value::U16(_) | Value::U32(_) | Value::U64(_) => {
+                return Err(Error::UnsupportedTypeOperation)
+            }
+        };
+        Ok(value)
+    }
+
+    /// Perform an addition operation.
+    ///
+    /// This operation requires matching types.
+    ///
+    /// This corresponds to the DWARF `DW_OP_plus` operation.
+    pub fn add(self, rhs: Value, addr_mask: u64) -> Result<Value> {
+        let value = match (self, rhs) {
+            (Value::Generic(v1), Value::Generic(v2)) => {
+                Value::Generic(v1.wrapping_add(v2) & addr_mask)
+            }
+            (Value::I8(v1), Value::I8(v2)) => Value::I8(v1.wrapping_add(v2)),
+            (Value::U8(v1), Value::U8(v2)) => Value::U8(v1.wrapping_add(v2)),
+            (Value::I16(v1), Value::I16(v2)) => Value::I16(v1.wrapping_add(v2)),
+            (Value::U16(v1), Value::U16(v2)) => Value::U16(v1.wrapping_add(v2)),
+            (Value::I32(v1), Value::I32(v2)) => Value::I32(v1.wrapping_add(v2)),
+            (Value::U32(v1), Value::U32(v2)) => Value::U32(v1.wrapping_add(v2)),
+            (Value::I64(v1), Value::I64(v2)) => Value::I64(v1.wrapping_add(v2)),
+            (Value::U64(v1), Value::U64(v2)) => Value::U64(v1.wrapping_add(v2)),
```

Comment:
I feel like all these operation implementations could use a macro or something to reduce boilerplate.

Reply:
What do you think of this: https://github.com/philipc/gimli/commit/d0b80cf09c4bd2ddd533ca84e31f3d668d5ef69d?

Reply:
Looks nice!